### ISSUE TITLE
[2.1.x] Function setVar() not working

### ISSUE THREAD

**Issue by zjhzxhz** [phalcon/cphalcon#11465]
I've upgrade to PHP 7.0.3 and Phalcon 2.1.x. But the code works well before upgrade can't works now.

Here's the code in controller:

```
class DefaultController extends BaseController {
    public function indexAction() {
        $this->view->setVar('appVersion', '2.0.0');
        $this->tag->prependTitle('Page Title');
    }
}
```

And here's the code in views:

```
<?php echo $appVersion; ?>
```

And I got an error message:

```
Notice: Undefined variable: appVersion in /home/xiehaozhe/Development/PHP/TestZilla/app/views/default/index.phtml on line 3
```

What's wrong with it?

**Comment by zjhzxhz** [comment]
Believe it or not, you can get the source code from [this repository](https://github.com/zjhzxhz/TestZilla).

**Comment by SidRoberts** [comment]
You need to do:

    <?php echo $this->appVersion; ?>

If you use Volt, you can write:

    {{ appVersion }}

**Comment by zjhzxhz** [comment]
No, I tried following statements:

```
<?php echo Phalcon\Tag::stylesheetLink("{$this->cdnUrl}/css/dashboard/style.css?ver={$this->appVersion}"); ?>
```

And I got following error message:

```
Notice: Access to undefined property appVersion in /home/xiehaozhe/Development/PHP/TestZilla/app/views/default/index.phtml on line 2
```

And the code works well in Phalcon 2.0.x with PHP 5.5 and PHP 5.6 :)

**Comment by zjhzxhz** [comment]
@sergeyklay 

I don't think @SidRoberts 's answer can solve this problem.
Please reopen this issue.

**Comment by sergeyklay** [comment]
@zjhzxhz 
For example:
* Set vars: https://github.com/phalcon/forum/blob/master/app/controllers/ControllerBase.php#L55-L61
* Using vars: https://github.com/phalcon/forum/blob/master/app/views/partials/statistics.volt#L1
* 2.1.x https://github.com/phalcon/forum/blob/master/.travis.yml#L19

**Comment by SidRoberts** [comment]
Actually, I think @zjhzxhz is right. I was thinking about DI services.

@zjhzxhz, could you post your View DI Service? Something like:
```php
$di->set("view", function(){
    // ...
});
```

**Comment by zjhzxhz** [comment]
Here's the View DI Service:

```
$di->set('view', function() use ($config) {
    $view = new View();
    $view->setViewsDir(APP_PATH . $config->application->viewDir);
    return $view;
});
```

And here's the reference: https://github.com/zjhzxhz/TestZilla/blob/master/app/config/services.php#L73

**Comment by zjhzxhz** [comment]
@sergeyklay

Here's my code:

```
$this->view->setVar('appVerion', APP_VERSION);
```

And here's the reference: https://github.com/zjhzxhz/TestZilla/blob/master/app/controllers/BaseController.php#L25-L37

I think it's the same as the example.

**Comment by SidRoberts** [comment]
"Version" has an S in it. :wink:

**Comment by sergeyklay** [comment]
:confused:

**Comment by zjhzxhz** [comment]
Yes, but in the view, I also forgot S.
And please DO KNOW that the code works well in Phalcon 2.0.x.

So I insist that it's your fault.

**Comment by zjhzxhz** [comment]
Another important thing to consider is that  all variables in views will report that error message, not only the variable appVersion.

**Comment by sergeyklay** [comment]
@zjhzxhz What do you think about that https://github.com/phalcon/cphalcon/blob/master/unit-tests/ViewTest.php#L64-L121

**Comment by dzubchik** [comment]
@zjhzxhz , I tried to reproduce this issue on identical configuration, but failed. If you are sure that setup is correct, please attach all files, that take part in rendering.

**Comment by zjhzxhz** [comment]
@dzubchik 
You can get all files here: https://github.com/zjhzxhz/TestZilla/archive/master.zip

**Comment by crocodile2u** [comment]
I can confirm this issue. Using Volt, Phalcon 2.1.x & PHP-7. Templates used to work fine with PHP5 & Phalcon 2.0, now they're all complaining about Undefined variables. Just as an idea as of what can be a reason, consider these lines in Volt.zep:

```
111                 /**
112                  * Export the variables the current symbol table
113                  */
114                 if typeof params == "array"     {
115                         for key, value in params {
116                                 let {key} = value;
117                         }
118                 }
```

let {key} = value - assignment to an undeclared variable. If you take a closer look, you will also find that if someone uses template varaibles with names "compiledTemplatePath" or "mustClean" - this will break the Volt::render(), as they will override zephir values with values from user-provided _viewParams. This code has not been changed in the new version of framework, but I suppose that maybe due to internal changes in Zephir the resulting behavior might have been altered.

**Comment by Jurigag** [comment]
As i know symbol tables was changed in php 7 - this is why this code does not work.

**Comment by crocodile2u** [comment]
Well this is weird:
1. I used to have this problem, accompanied by "memory limit exhausted" on every call to Model::find/findFirst/count(). 
2. I discovered that I forgot to add php-imagick extension to PHP-7. I actually did not think that adding it would change anything in the Phalcon-based project - since the code crashed in other places than those that relied on Imagick.
3. I installed Imagick
4. **The crashes stopped**.

I mean, they stopped without me recompiling phalcon - just adding imagick + php-fpm restart. Setting view variables now works fine without any workarounds etc. This is good news from the first sight. But this also means for me that everything could crash again for no apparent reason at any time... 

Unfortunately, I'm not able to provide any test script that could illustrate this behavior ((

P. S. I updated the branch, compiled phalcon again (zephir fullclean && zephir build). So far my project works fine. Luckily, this is just a personal photoblog hosted at home on Raspberry Pi 3, so nothing too much critical )))
